### DISCUSSION TITLE
missing data in  Subscription

### DISCUSSION THREAD

**Discussion by xyzakdkd** [FreeOpcUa/opcua-asyncio#1643] [Q&A] [answered]
Hello, After establishing a connection to an OPC UA server using subscription, data retrieval appears to be functioning smoothly. However,  when i compare the data obtained from my script with that data using the (UaExpert Software) Datalogger. Notably by running both for 3 hours,  i get missing data on my client script. **In the table you can see that the bold value is missing**

### Data from the DataLogger
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/walee/AppData/Local/Temp/msohtmlclip1/02/clip.htm">
<link rel=File-List
href="file:///C:/Users/walee/AppData/Local/Temp/msohtmlclip1/02/clip_filelist.xml">
</head>

<body link="#467886" vlink="#96607D">

x | xxx | value | xxx | source_timestamp | server_timestamp
-- | -- | -- | -- | -- | --
1 | 0x000b | 72.5 | 0x00000000 | 2024-05-13T13:59:07.753Z | 2024-05-13T13:59:07.784Z
2 | 0x000b | 72.40000153 | 0x00000000 | 2024-05-13T13:59:08.999Z | 2024-05-13T13:59:09.084Z
3 | 0x000b | 72.30000305 | 0x00000000 | 2024-05-13T13:59:10.568Z | 2024-05-13T13:59:10.584Z
4 | 0x000b | 72.19999695 | 0x00000000 | 2024-05-13T13:59:12.001Z | 2024-05-13T13:59:12.084Z
5 | 0x000b | 72 | 0x00000000 | 2024-05-13T13:59:15.189Z | 2024-05-13T13:59:15.284Z
6 | 0x000b | 71.80000305 | 0x00000000 | 2024-05-13T13:59:15.690Z | 2024-05-13T13:59:15.785Z
7 | 0x000b | 71.69999695 | 0x00000000 | 2024-05-13T13:59:16.065Z | 2024-05-13T13:59:16.087Z
8 | 0x000b | 71.5 | 0x00000000 | 2024-05-13T13:59:16.627Z | 2024-05-13T13:59:16.685Z
**9** | **0x000b** | **71.80000305** | **0x00000000** | **2024-05-13T13:59:17.002Z** | **2024-05-13T13:59:17.085Z**
10 | 0x000b | 71.90000153 | 0x00000000 | 2024-05-13T13:59:17.004Z | 2024-05-13T13:59:17.087Z
11 | 0x000b | 72.09999847 | 0x00000000 | 2024-05-13T13:59:17.252Z | 2024-05-13T13:59:17.285Z
12 | 0x000b | 72.40000153 | 0x00000000 | 2024-05-13T13:59:17.502Z | 2024-05-13T13:59:17.585Z
13 | 0x000b | 72.59999847 | 0x00000000 | 2024-05-13T13:59:17.753Z | 2024-05-13T13:59:17.784Z
14 | 0x000b | 72.80000305 | 0x00000000 | 2024-05-13T13:59:17.876Z | 2024-05-13T13:59:17.886Z
15 | 0x000b | 73.09999847 | 0x00000000 | 2024-05-13T13:59:18.003Z | 2024-05-13T13:59:18.088Z
16 | 0x000b | 73.30000305 | 0x00000000 | 2024-05-13T13:59:18.375Z | 2024-05-13T13:59:18.387Z
</body>
</html>

### Data from my script
![Screenshot (536)](https://github.com/FreeOpcUa/opcua-asyncio/assets/147828509/3fc57589-ac68-4f92-bd82-40859fa82682)

**Comment by AndreasHeine** [comment]
you need to post the subscription and monitored item settings!

**Reply by xyzakdkd** [reply]
can you please give me an example

### Subscription part

```
async def create_subscription(client, handler):
    try:
        subscription = await client.create_subscription(500, handler)
        handler.subscription = subscription
        for node_id in nodes_list:
            node = client.get_node(node_id)
            await subscription.subscribe_data_change(node, queuesize=1)
    except (ConnectionError, ua.UaError):
        _logger.warning("Error creating subscription")
```

**Reply by AndreasHeine** [reply]
might be the queuesize!

**Reply by xyzakdkd** [reply]
i change it to 0 and still get missing data

**Reply by AndreasHeine** [reply, marked as answer]
what are the settings of UAExpert? tbh dataloss is very unlikely due to the fact that the server fill the queue and the client recv all queued changes within one response. i bet its just a setting thing!

**Reply by xyzakdkd** [reply]
I try it to change the setting on the UAExpert. are this the same as the default setting in asyncua? all i need to add is the Sampling interval

### my code

```
async def create_subscription(client, handler):
    try:
        subscription = await client.create_subscription(100, handler)
        handler.subscription = subscription
        for node_id in nodes_list:
            node = client.get_node(node_id)
            await subscription.subscribe_data_change(node, sampling_interval = 250)
    except (ConnectionError, ua.UaError):
        _logger.warning("Error creating subscription")
```

![Screenshot (537)](https://github.com/FreeOpcUa/opcua-asyncio/assets/147828509/2550c25c-a6bb-4019-b6db-377305f04b9d)

**Reply by xyzakdkd** [reply]
Thanks @AndreasHeine you were right, it was just a setting thing

**Reply by AndreasHeine** [reply]
@xyzakdkd glad i could help!

**Reply by xyzakdkd** [reply]
Hello @AndreasHeine i was trying the whole week to test it, it sometime give the same data as the UA-Expert but in many time i get more data in the UA-Expert. Where could be the problem. or  could be a problem with the server.   

### In the code:
```
async def create_subscription(client, handler):
    try:
        subscription = await client.create_subscription(200, handler)
        handler.subscription = subscription
        for node_path in x:
            node = client.get_node(node_path)
            await subscription.subscribe_data_change(node,  queuesize=10, sampling_interval=250)
    except (ConnectionError, ua.UaError):
        _logger.warning("Error creating subscription")
```

**CreateSubscriptionResult(SubscriptionId=2156708, RevisedPublishingInterval=200.0, RevisedLifetimeCount=10000, RevisedMaxKeepAliveCount=3000)**

### UA-Expert Setting 
![Screenshot (540)](https://github.com/FreeOpcUa/opcua-asyncio/assets/28221354/3ae7f5d4-71e6-4726-a3fb-dc17e918669f)

**Reply by AndreasHeine** [reply]
try longer publishinterval like 1s and queue  100 just to compare... sampling can stay 250ms

generally speaking the server queues changes and the client recvs them in batches... if there is nothing in the debug logs of asyncua then it is possible that it is server but try the above settings first

**Reply by xyzakdkd** [reply]
Sorry for the late response. I was testing, and you were right. i change the queue and everything worked. thanks
